receiver/sapmreceiver/trace_receiver.go

```diff
@@ -0,0 +1,173 @@
+	// receiverFn is a discrete function passed to sapm-proto's http.HanlderFunc.  This receiverFn actually processes the
+	// sapm and passes it on to the next trace consumer
+	receiverFn := func(ctx context.Context, sapm *splunksapm.PostSpansRequest, err error) error {
+		rCtx := observability.ContextWithReceiverName(ctx, "sapm")
+
+		// process the sapm
+		if sapm != nil {
+			for _, batch := range sapm.Batches {
+				// convert the jager batches to OCProto
+				td, err := jaegertranslator.ProtoBatchToOCProto(*batch)
+				if err != nil {
+					return err
+				}
+				td.SourceFormat = "sapm"
+
+				// pass the trace data to the next consumer
+				sr.nextConsumer.ConsumeTraceData(rCtx, td)
+
+				// We MUST unconditionally record metrics from this reception.
+				observability.RecordMetricsForTraceReceiver(rCtx, len(batch.Spans), len(batch.Spans)-len(td.Spans))
+			}
+		}
+
+		return nil
+	}
+
+	// get a handler function from the sapm-proto repo and pass it the receiverFn
+	sapmHandlerFn := sapmhandler.NewTraceHandlerV2(receiverFn)
+
+	// return an http.HandlerFunc that traces the request and passes it through the sapm-proto handler
+	return func(rw http.ResponseWriter, req *http.Request) {
+
+		// trace this request
+		_, span := trace.StartSpan(sr.defaultCtx, traceSource)
+		sapmHandlerFn(rw, req)
+		span.End()
+
+		return
+	}
+}
+
+// TraceSource implements receiver.TraceReceiver.TraceSource() and returns a tag describing the source format
+func (sr *sapmReceiver) TraceSource() string {
+	return traceSource
+}
+
+// StartTraceReception starts the sapmReceiver's server
+func (sr *sapmReceiver) StartTraceReception(host receiver.Host) error {
+	sr.mu.Lock()
+	defer sr.mu.Unlock()
+
+	var err = oterr.ErrAlreadyStarted
+	sr.startOnce.Do(func() {
+		err = nil
+		// set up the listener
+		ln, lerr := net.Listen("tcp", sr.config.Endpoint)
+		if lerr != nil {
+			err = fmt.Errorf("failed to bind to address %s: %v", sr.config.Endpoint, lerr)
+		}
+
+		// use gorilla mux to create a router/handler
+		nr := mux.NewRouter()
+		nr.HandleFunc(sapmhandler.TraceEndpointV2, sr.HTTPHandlerFunc())
+
+		// create a server with the handler
+		sr.server = &http.Server{Handler: nr}
+
+		// run the server on a routine
+		go func() {
+			host.ReportFatalError(sr.server.Serve(ln))
+		}()
+	})
+	return err
+}
+
+// StopTraceRetention stops the the sapmReceiver's server
+func (sr *sapmReceiver) StopTraceReception() error {
+	sr.mu.Lock()
+	defer sr.mu.Unlock()
+
+	var err = oterr.ErrAlreadyStopped
+	sr.stopOnce.Do(func() {
+		var errs []error
+
+		if sr.server != nil {
+			if serr := sr.server.Close(); serr != nil {
+				errs = append(errs, serr)
+			}
+			sr.server = nil
+		}
+		if len(errs) == 0 {
+			err = nil
+			return
+		}
+		// Otherwise combine all these errors
+		buf := new(bytes.Buffer)
+		for _, err := range errs {
+			fmt.Fprintf(buf, "%s\n", err.Error())
+		}
+		err = errors.New(buf.String())
```

Comment:
Otel has a package to combine multiple errors: https://github.com/open-telemetry/opentelemetry-collector/blob/master/oterr/errors.go

Reply:
cool, I copied that from the jaeger receiver.  I'll update to use that.

Reply:
didn't actually have multiple errors any way.  It was a bad copy/paste.  I've removed it entirely.